### ISSUE TITLE
Error: input and output filenames must not be matched

### ISSUE THREAD

**Issue by PaulBurridge-Nasstar** [jazzband/pip-tools#1823]
Error: input and output filenames must not be matched

#### Environment Versions

1. OS Type: Windows 11
2. Python version: 3.9.17
3. pip version: 23.0.1
4. pip-tools version: 6.12.2

#### Steps to replicate

Using command as per comments in and existing compiled requirements.txt after adding a new dependency in file.

```
#
# This file is autogenerated by pip-compile with Python 3.9
# by the following command:
#
#    pip-compile --annotation-style=line ./bin/app/requirements.txt
#
click==8.1.3              # via flask
flask==2.2.2              # via flask-failsafe
flask-failsafe==0.2       # via -r ./bin/app/requirements.txt
importlib-metadata==6.0.0  # via flask
itsdangerous==2.1.2       # via flask
jinja2==3.1.2             # via flask
markupsafe==2.1.2         # via jinja2, werkzeug
werkzeug==2.2.2           # via flask
zipp==3.11.0              # via importlib-metadata
gunicorn # New dependecy
```

Executed

`pip-compile --annotation-style=line ./bin/app/requirements.txt`

#### Expected result

File recompiled with new dependency fixed and added dependencies.

#### Actual result

`Error: input and output filenames must not be matched`

#### Notes

This appears to be caused by changes in this PR https://github.com/jazzband/pip-tools/pull/1787 "Raise error if input and output filenames are matched"

I can understand thinking behind the change but there is little risk of updating file in place when using version control, it means that to add a new dependency I have to copy the file first and add an --output-file parameter, then delete the input file, or manage two files in the repository going forward.

It is also at odds with the deriving of output file name from input as specified as a option in help:

```
  -o, --output-file FILENAME      Output file name. Required if more than one
                                  input file is given. Will be derived from
                                  input file otherwise.
```

It breaks backwards compatibility withe the command list in comments in the output file.

Suggestion either remove this check or add a flag such as `--allow-overwrite` to suppress the check.

**Comment by AndydeCleyre** [comment]
The idea of the tool is to keep a `requirements.in` file with your explicit, top level dependencies, versioned as loosely as possible, then generate a fully version locked, deep dependency tree/list as `requirements.txt`.

If you then remove the `.in` file and compile using the locked `.txt` as a source, you lose the information of which dependencies are explicit/top level, and you lose the possibility of the versions getting updated, because they're already locked in exactly.

It is very unlikely that is any user's intent.

**Comment by PaulBurridge-Nasstar** [comment]
It is not necessary to use a separate file to tell the top level dependencies as below: 

`# via -r requirements.txt` - top level
`# via -r requirements.txt, flask` - dependency of flask

We have many developers who are not used to using pip-compile who simply seek out a `requirements.txt` file to add dependencies to, we would prefer this continued to be possible going forward. 

The tool has worked fine using a single file before this change and if the "idea of this tool" is not to ever use a single file why does the help for `--output-file` specify `Required if more than one input file is given.  Will be derived from input file otherwise`?

**Comment by PaulBurridge-Nasstar** [comment]
Also to update versions we simply delete those that are not top level from the file, remove any relevant fixed version then recompile.

**Comment by AndydeCleyre** [comment]
> It is not necessary to use a separate file to tell the top level dependencies as below:

The way you're doing it, if an item is both a top level requirement and required by another package, you will have to remember that fact, it won't be reflected in your output.

> ... if the "idea of this tool" is not to ever use a single file why does the help for `--output-file` specify `Required if more than one input file is given.  Will be derived from input file otherwise`?

That's two files, not a single file. The input file is the `.in`, and the output file is the `.txt`.

**Comment by AndydeCleyre** [comment]
If you need to specify a version range, that will also be lost during compilation, and you'll need to remember that too.

**Comment by PaulBurridge-Nasstar** [comment]
I have always run without specifying an output file and before the change it did overwrite the existing file which suggests that is a valid workflow by design. (see below)

```
#
# This file is autogenerated by pip-compile with Python 3.9
# by the following command:
#
#    pip-compile --annotation-style=line ./bin/app/requirements.txt
#
```

**Comment by atugushev** [comment]
> he message "Raise error if input and output filenames are matched" doesn't communicate the actual problem

I agree the error message could be better. I'd gladly review and merge a PR 👍🏻
